### PR TITLE
Task Scheduler Configuration

### PR DESCRIPTION
**Pull request by eskyuu**
This aims to re-work the way different task schedulers are configured.  It sets a single location where the scheduling method is selected instead of having different enable/disable toggles.

Once completed, it would technically be possible for users to leave the default crontabs and service daemon running, and use the UI to select the polling method.

image

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [x] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [x] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.
- [x] If my Pull Request makes discovery/polling/yaml changes, I have added/updated [test data](https://docs.librenms.org/Developing/os/Test-Units/).

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

File: alerts.php [hunk 1]
```diff
@@ -33,12 +33,20 @@
...
 require __DIR__ . '/includes/init.php';
 
-$options = getopt('d::');
+$options = getopt('fd::');
 
 if (Debug::set(isset($options['d']))) {
     echo "DEBUG!\n";
 }
 
+$scheduler = Config::get('schedule_type.discovery');
```
Comment (murrant) [inline alerts.php:42]: Typo here

**Comment by murrant** [comment]
I think the setting structure is correct.

It seems like this affects manual calls too.

What is the -f option for?

**Comment by eskyuu** [comment]
Yes, the setting affect manual calls once someone changes the setting to something other than "default"

The -f option is to "force" the command to run even if it's not configured to do so.  It is there so the service daemon can still run the php scripts even if the cron job is disabled.  I will add something to the debug log saying to use -f to force the command to run.

File: alerts.php (hunk 1 above)
Comment (eskyuu) [reply alerts.php]: Thanks - fixed

**Comment by murrant** [comment]
What if we invert the option a bit and only check when it is specified?

So, like -s dispatcher and -s wrapper (and -s cron, but it would only apply to new installs as no one updates cron entries)

**Comment by murrant** [comment]
I think I was focusing on polling/discovery, but I see those aren't much the problem...

But what if we implemented the commands as Laravel commands and only add the check to the legacy .php scripts?  This would be a good compromise I think.   No one should have lnms commands in their cron directly (except the scheduler).

**Comment by eskyuu** [comment]
I agree that implementing the commands as Laravel commands would be a good way to avoid the need to check for people running manually, and it would mean that the wrapper and service daemon can run the lnms command without needing the "-f" flag.

Having said that, I think that creating new jobs, commands and schedules for these tasks will become easy improvements once we have this config option and the polling scheduled task implemented.

I think the "-f" option for anyone who actively changes this new config option and then wants to manually run the command is acceptable to avoid expanding the scope of this PR (or making the Laravel scheduler for the poller dependent on another PR).

File: doc/Extensions/Services.md [hunk 2]
```diff
@@ -210,7 +210,7 @@ These settings are related and should be investigated and set accordingly.
 The below values are not defaults or recommended.
 
 ```php
-$config['service_poller_enabled']           = true;
```
Comment (murrant) [inline doc/Extensions/Services.md:213]: Hmm, this section of the docs seems very unrelated to services.  Perhaps it should be trimmed down.

Comment (murrant) [reply doc/Extensions/Services.md:213]: Also, could swap to lnms config:set

Comment (eskyuu) [reply doc/Extensions/Services.md:213]: In this case I assumed that the documentation was left this way to be able to override this setting on a per node basis.

I think that some of the documentation around polling config including distributed polling may need a larger review.  For example the scaling document for distributed polling says that redis is preferred for locking, but the web UI, sample configuration and validation scripts only show how to configure memcached.  I think the service daemon may only use redis for queue management (but its documentation mentions memcached), while the cron wrappers only use memcache for device level locking.  There is also the mention of poller config in the services page above.

Comment (murrant) [reply doc/Extensions/Services.md:213]: I would say to leave it, but I think the change only adds to the existing confusion.

**Review by murrant — APPROVED** [review]

**Review by murrant — DISMISSED** [review]

**Comment by laf** [comment]
I haven't read through the code as murrant has already approved it but I just want to check:

1. What does the legacy option translate to, cron or the dispatcher?
2. Is this a breaking change at all regardless of the way people run LibreNMS?

Also needs a rebase @eskyuu

**Comment by murrant** [comment]
Legacy = unrestricted polling anything is allowed.
The other settings only allow the specified polling type.
By making the defaults legacy, it will not be a breaking change.

We struggled finding a good name for the legacy setting, open to suggestions to make it easier to understand.

**Comment by eskyuu** [comment]
> I haven't read through the code as murrant has already approved it but I just want to check:
> 
>     1. What does the legacy option translate to, cron or the dispatcher?
> 
>     2. Is this a breaking change at all regardless of the way people run LibreNMS?
> 
> 
> Also needs a rebase @eskyuu

To answer both questions - Legacy translates to cron and/or the dispatcher based on the existing behaviour, where cron is controlled by enabling/disabling the crontab entry, and dispatcher is enabled by both configuring the service and enabling the appropriate config option on a per-job basis.  The legacy option is there to ensure we don't break existing deployments.

This change aims to make it easier to add new options in future, and also remove the need to disable cron jobs when changing schedulers.

**Comment by eskyuu** [comment]
Is it possible to get this merged even though the persistent workers has stalled.  I still think this code is a good step forward for configuring the existing poller options.

**Comment by laf** [comment]
> Is it possible to get this merged even though the persistent workers has stalled. I still think this code is a good step forward for configuring the existing poller options.

I'm good with it personally. I'll give it a proper test tomorrow to be sure.

On a name for legacy, what about 'Inherited' if the option means we detected the currently used one and pulled it in?

**Review by murrant — APPROVED** [review]

**Comment by murrant** [comment]
I plan to get back to persistent workers.

**Pull request merged**